egs/wsj/s5/steps/conf/train_calibration.sh

```diff
@@ -0,0 +1,120 @@
+# - log of frames per character ratio
+# (- categorical distribution of 'lang/words.txt', DISABLED)
+
+# begin configuration section.
+cmd=
+lmwt=12
+decode_mbr=true
+word_min_count=10 # Minimum word-count for single-word category,
+normalizer=0.0025 # L2 regularization constant,
+stage=0
+# end configuration section.
+
+[ -f ./path.sh ] && . ./path.sh
+. parse_options.sh || exit 1;
+
+if [ $# -ne 7 ]; then
+  echo "Usage: $0 [opts] <data-dir> <lang-dir|graph-dir> <word-filter> <word-length> <unigrams> <decode-dir> <calibration-dir>"
+  echo " Options:"
+  echo "    --cmd (run.pl|queue.pl...)      # specify how to run the sub-processes."
+  echo "    --lmwt <int>                    # scaling for confidence extraction"
+  echo "    --decode-mbr <bool>             # use Minimum Bayes Risk decoding"
+  echo "    --grep-filter <str>             # remove words from calibration targets"
+  exit 1;
+fi
+
+set -euxo pipefail
+
+data=$1
+lang=$2 # Note: may be graph directory not lang directory, but has the necessary stuff copied.
+word_filter=$3
+word_length=$4
+unigrams=$5
+latdir=$6
+dir=$7
+
+model=$latdir/../final.mdl # assume model one level up from decoding dir.
+
+for f in $data/text $lang/words.txt $word_filter $word_length $unigrams $latdir/lat.1.gz; do
+  [ ! -f $f ] && echo "$0: Missing file $f" && exit 1
+done
+[ -z "$cmd" ] && echo "$0: Missing --cmd '...'" && exit 1
+
+[ -d $dir/log ] || mkdir -p $dir/log
+nj=$(cat $latdir/num_jobs)
+
+# Store the setup,
+echo $lmwt >$dir/lmwt
+echo $decode_mbr >$dir/decode_mbr
+cp $word_filter $dir/word_filter
+cp $word_length $dir/word_length
+cp $unigrams $dir/unigrams
+
+# Create the ctm with raw confidences,
+# - we keep the timing relative to the utterance,
+if [ $stage -le 0 ]; then
+  $cmd JOB=1:$nj $dir/log/get_ctm.JOB.log \
+    lattice-scale --inv-acoustic-scale=$lmwt "ark:gunzip -c $latdir/lat.JOB.gz|" ark:- \| \
+    lattice-limit-depth ark:- ark:- \| \
+    lattice-push --push-strings=false ark:- ark:- \| \
+    lattice-align-words-lexicon --max-expand=10.0 \
+     $lang/phones/align_lexicon.int $model ark:- ark:- \| \
+    lattice-to-ctm-conf --decode-mbr=$decode_mbr ark:- - \| \
+    utils/int2sym.pl -f 5 $lang/words.txt \
+    '>' $dir/JOB.ctm
+  # Merge and clean,
+  set +x; for ((n=1; n<=nj; n++)); do cat $dir/${n}.ctm; done > $dir/ctm
+  rm $dir/*.ctm; set -x
+fi
+
+# Get evaluation of the 'ctm' using the 'text' reference,
+if [ $stage -le 1 ]; then
+  steps/conf/convert_ctm_to_tra.py $dir/ctm - | \
+  align-text --special-symbol="<eps>" ark:$data/text ark:- ark,t:- | \
+  utils/scoring/wer_per_utt_details.pl --special-symbol "<eps>" \
+  >$dir/align_text 
+  # Append alignment to ctm,
+  steps/conf/append_eval_to_ctm.py $dir/align_text $dir/ctm $dir/ctm_aligned
+fi
+
+# Prepare word-categories (based on wotd frequencies in 'ctm'),
+steps/conf/prepare_word_categories.py --min-count $word_min_count $lang/words.txt $dir/ctm $dir/word_categories
+
```

Comment:
Regarding whether to get the word categories from 'dev' or train:
  - How about putting it in the outer level, user-controlled script, and let them run the script that gets them?  That way, in the normal case they can get them on the dev set; but they can choose which set to get them from?  In general I think it's nicer to have as many of the word-specific features as possible come from the user rather than be hard-coded in the script that does the confidence calibration.